crates/wit-parser/src/decoding.rs

```diff
@@ -204,25 +342,23 @@ impl ComponentInfo {
         }
 
         let pkg = pkg.ok_or_else(|| anyhow!("no exported component type found"))?;
-        let (mut resolve, package) = decoder.finish(pkg);
+        let (mut resolve, packages) = decoder.finish(&mut [pkg]);
         if let Some(package_metadata) = &self.package_metadata {
-            package_metadata.inject(&mut resolve, package)?;
+            package_metadata.inject(&mut resolve, packages[0])?;
         }
-        Ok((resolve, package))
+        Ok((resolve, packages[0]))
     }
 
-    fn decode_wit_v2_package(&self) -> Result<(Resolve, PackageId)> {
+    fn decode_wit_v2_packages(&self) -> Result<(Resolve, Vec<PackageId>)> {
```

Comment:
I'm a bit lost in the changes to this method, can you explain why so many changes were needed? My (probably) naive understanding is that this could work by decoding each sub-component if present in the exact same way they're all decoded today, and then at the end the `Resolve` values are all merged together. This looks like it's lazily inserting packages and/or using various other bits of logic I don't fully understand. In other words I'd expect that there wouldn't be much change in general and instead decoding would decode each sub-component in a loop and then merge everything afterwards.

Reply:
So when parsing the binary today, the only metadata captured in `ComponentInfo` is for the "implicit" package at the root, and the decoding is based on the metadata there.  I don't think we have the ability to decode each subcomponent as they're encountered, just based on how `ComponentInfo::from_reader` parses the binary today, I think it has go through the entire binary, and I don't have any great ideas about how to change that.  I didn't realize initially that I had the `merge` method on `Resolve` available to me though.  Since you're comment, I updated to decode each `ExplicitComponentInfo` separately, and then merge each of the resolves together.

I think most of the changes at this point are just clearing out the `PackageFields` between decoding each `ExplicitPackageInfo` , or else interfaces/worlds carry over from when package to the next.  I can look into seeing if there is a way to make this a bit tidier though/potentially create a new `PackageFields` per package rather than reusing and clearing the same one as we iterate through packages.

Reply:
Ah ok that makes sense. I agree the best way to go about this is to modify how `ComponentInfo` is collected and processed.